internal/cmd/controller/reconciler/cluster_controller.go

```diff
@@ -337,3 +346,42 @@ func (r *ClusterReconciler) mapBundleDeploymentToCluster(ctx context.Context, a
 		},
 	}}
 }
+
+// ensureFinalizer adds a finalizer to a recently created cluster.
+func (r *ClusterReconciler) ensureFinalizer(ctx context.Context, cluster *fleet.Cluster) error {
+	if controllerutil.ContainsFinalizer(cluster, finalize.ClusterFinalizer) {
+
+		return nil
+	}
+
+	controllerutil.AddFinalizer(cluster, finalize.ClusterFinalizer)
+
+	return r.Update(ctx, cluster)
+}
+
+// handleDelete runs cleanup the namespace associated to a Cluster,
+// finally removing the finalizer to unblock the deletion of the object from kubernetes.
```

Comment:
Cleaning up the namespaces should also be done in the [cleanup controller](https://github.com/rancher/fleet/blob/main/internal/cmd/controller/cleanup/controllers/cleanup/controller.go#L93-L100).

I am wondering why this did not work out but we potential could get rid of this code path since the new approach should be more reliable.